### DISCUSSION TITLE
shutdowntimer on headlesspi.py

### DISCUSSION THREAD

**Discussion by ricardojsmarques** [GeekFunkLabs/fluidpatcher#3] [General]
Hi!
Great job with the fluidpatcher code, it is allowing me to use my RPI3 again and make my Oxygen 61 controller independent!

I have noticed that when I'm playing that sometimes the PI shuts down (gracefully).
Upon investigation, I found that there is a shutdowntimer variable and logic that shuts down the PI. 

What is the use for it and the logic behind it?

**Comment by albedozero** [comment]
Thanks! If you hold down one of the patch increment/decrement buttons for 7 seconds, it shuts down the Pi gracefully, so that you don't have to unplug the Pi and possibly corrupt your SD card - that's what all the `shutdowntimer` logic is for. I just added this 4 days ago in commit [7536828](https://github.com/albedozero/fluidpatcher/commit/75368289931f7a2e8e04a7f61b82b8ea9585bf99) and haven't had a chance to document it yet. If you don't want this to happen you can just comment out [this line](https://github.com/albedozero/fluidpatcher/blob/75368289931f7a2e8e04a7f61b82b8ea9585bf99/headlesspi.py#L143) in headlesspi.py and the timer will never get started.

If you don't think this was causing your issue, I have had the Pi shut down at times when there was a bug I needed to fix. Since the script runs at startup as a `systemd` process it will try to recover if it breaks, and if it keeps failing it will exit, which might seem like the Pi is shutting down. If you log in remotely via `ssh` you can run the script interactively by entering
```
sudo systemctl stop squishbox.service
./headlesspi.py --interactive
```
That could give you a better idea what's going on if something breaks.

**Reply by ricardojsmarques** [reply]
Now I understand why it was shutting down randomly, :D
I was using MIDI CC 23, on a knob, and while trying to search for a patch, it ended up shutting down. 

I think it would be less "error prone" if there was the possibility to press both INC_PATCH and DEC_PATCH buttons for 7 seconds to initiate the shutdown so one doesn't shutdown the PI by accident.

If also there was the possibility to maybe emit a very short, low frequeny beep when the pi is ready to play would be nice :)

**Reply by albedozero** [reply]
Hmm - SELECT_PATCH shouldn't trigger a shutdown, only INC_PATCH or DEC_PATCH. However, it does occur to me that not all synths send a CC value of zero when you release a button, and maybe some don't do it consistently - without which the shutdown timer doesn't get canceled. You could see if your controller is doing this by entering `aconnect -iol` on the command line to find your midi controller's port number (probably something like `20:0`), then enter `aseqdump -p 20:0` to see what MIDI messages it's sending. Might have to think of a better way to code this.. I'd thought about the pressing INC_PATCH and DEC_PATCH together, but didn't want to force people to assign both those buttons.

The Pi does blink the green activity LED 5 times when it's ready to play, and it should start blinking the red LED when you're about to shut it down.
